main.go

```diff
@@ -1024,48 +1031,115 @@ type funcFullName = string
 // knownReflectAPIs is a static record of what std APIs use reflection on their
 // parameters, so we can avoid obfuscating types used with them.
 //
-// For now, this table is manually maintained, until we automatically detect and
-// propagate the uses of reflect in std and third party APIs.
-// If we end up maintaining this table for a long time, we should probably
-// improve the process via either code generation or sanity checks.
-//
-// TODO: record which parameters get used with reflection, as it's often not all
-// of them.
-//
 // TODO: we're not including fmt.Printf, as it would have many false positives,
 // unless we were smart enough to detect which arguments get used as %#v or %T.
-//
-// TODO: this should also support third party APIs; see
-// https://github.com/burrowers/garble/issues/162
-var knownReflectAPIs = map[funcFullName]bool{
-	"reflect.TypeOf":  true,
-	"reflect.ValueOf": true,
-
-	"encoding/json.Marshal":           true,
-	"encoding/json.MarshalIndent":     true,
-	"encoding/json.Unmarshal":         true,
-	"(*encoding/json.Decoder).Decode": true,
-	"(*encoding/json.Encoder).Encode": true,
-
-	"encoding/gob.Register":          true,
-	"encoding/gob.RegisterName":      true,
-	"(*encoding/gob.Decoder).Decode": true,
-	"(*encoding/gob.Encoder).Encode": true,
-
-	"encoding/xml.Marshal":                  true,
-	"encoding/xml.MarshalIndent":            true,
-	"encoding/xml.Unmarshal":                true,
-	"(*encoding/xml.Decoder).Decode":        true,
-	"(*encoding/xml.Encoder).Encode":        true,
-	"(*encoding/xml.Decoder).DecodeElement": true,
-	"(*encoding/xml.Encoder).EncodeElement": true,
-
-	"(*text/template.Template).Execute": true,
-
-	"(*html/template.Template).Execute": true,
-
-	"net/rpc.Register":     true,
-	"net/rpc.RegisterName": true,
+var knownReflectAPIs = map[funcFullName][]int{
+	"reflect.TypeOf":  {0},
+	"reflect.ValueOf": {0},
+}
+
+func loadKnownReflectAPIs() error {
+	dir := filepath.Join(os.TempDir(), "garble")
+	filename := filepath.Join(dir, "reflect.json")
+
+	data, err := os.ReadFile(filename)
+	if err != nil || len(data) == 0 {
+		return nil
+	}
+
+	return json.Unmarshal(data, &knownReflectAPIs)
+}
+
+func saveKnownReflectAPIs() error {
```

Comment:
So each compile process stores a fixed reflect.json file? Isn't this racy?

Reply:
Does go build compile multiple packages in parallel?

Reply:
Yep. Builds would be terribly slow if compiles were serialized.

Reply:
This shouldn't be a problem for reading, because all dependencies need to be compiled before.
Writing would need to lock the file read the current state from disk and merge it with the in memory state before writing it to disk again.

Reply:
I still think this is wrong for multiple reasons:

1) You can't assume that /tmp will survive just like Go's build cache. For example, I use tmpfs for /tmp, so if I reboot, my Go build cache is kept, but my /tmp is empty. So if I do an incremental garble build after a reboot, your json data will be missing, and won't be rebuilt because an incremental build will not rebuild some of the packages.

2) If there are multiple garble builds happening concurrently (or even one after another?), they will share this json file. And since different builds may have different versions of modules and packages, they cannot safely share this reflect information.

The only safe place to store this info for reuse is in the build cache. I shared some notes on Slack on how to do that and why.